core/src/Cuda/Kokkos_Cuda_Parallel.hpp

```diff
@@ -1099,6 +1112,124 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
+              typename std::enable_if<!std::is_same<TagType_, void>::value,
+                                      bool>::type = true>
+    KOKKOS_FUNCTION value_type operator()(index_type i) const {
+      // value_type val = InitValueType();
+      value_type val = (value_type)0;  // for now, assuming no init value
+      f(TagType(), i, val);
+      return val;
+    }
+  };
+  */
+
+  static constexpr bool is_thrust_possible =
+      !ReduceFunctorHasInit<FunctorType>::value &&
+      !ReduceFunctorHasJoin<FunctorType>::value &&
+      !ReduceFunctorHasFinal<FunctorType>::value &&
+      !Policy::is_graph_kernel::value &&
+      std::is_same<ReducerType, InvalidType>::value;
+
+  template <class TagType>
+  struct ThrustFunctorWrapper {
+    const FunctorType f;
+    value_type init;
+
+    KOKKOS_FUNCTION
+    ThrustFunctorWrapper(const FunctorType op, value_type init)
+        : f(std::move(op)), init(std::move(init)){};
+
+    template <class TagType_                      = TagType,
+              typename std::enable_if<std::is_same<TagType_, void>::value,
+                                      bool>::type = true>
+    KOKKOS_FUNCTION value_type operator()(index_type i) const {
+      value_type val = init;
+      f(i, val);
+      return val;
+    }
+
+    template <class TagType_                      = TagType,
+              typename std::enable_if<!std::is_same<TagType_, void>::value,
+                                      bool>::type = true>
+    KOKKOS_FUNCTION value_type operator()(index_type i) const {
+      value_type val = init;
+      f(TagType(), i, val);
+      return val;
+    }
+  };
+
+  template <bool try_thrust>
+  inline std::enable_if_t<!try_thrust, bool> thrust_execute(bool) {
+    return false;
+  }
+  template <bool try_thrust>
+  inline std::enable_if_t<try_thrust, bool> thrust_execute(
+      bool thrust_runtime_possible) {
+    if (!thrust_runtime_possible) {
+      return false;
+    }
+    printf("using CUDA Thrust\n");
+
+    thrust::counting_iterator<index_type> temp_iter_d(m_policy.begin());
+
+    thrust::counting_iterator<index_type> temp_iter_end_d(m_policy.end());
+
+    // Used for debugging
+    // printf("m_policy.begin(): %d\n", m_policy.begin());
+    // printf("m_policy.end(): %d\n", m_policy.end());
+
+    value_type sum;
+
+    // if want to check for default constructible for complex types
+    // also needs an overloaded operator+ outside the struct scope
+    /*
+    if (!std::is_fundamental<value_type>::value &&
+        !std::is_default_constructible<value_type>::value){
+        printf("must be default constructible\n");
+    } else {
+        value_type init;
+    }
+    */
+    value_type init{};
+
+    ThrustFunctorWrapper<WorkTag> t_op(m_functor, init);
+
+    m_policy.space().fence();
+
+    sum = thrust::transform_reduce(thrust::device, temp_iter_d, temp_iter_end_d,
+                                   t_op, t_op.init, thrust::plus<value_type>());
```

Comment:
Can't this be
```suggestion
    sum = thrust::transform_reduce(
      thrust::device, temp_iter_d, temp_iter_end_d,
      t_op, t_op.init, 
      KOKKOS_LAMBDA(const value_type& lhs, const value_type& rhs) {
        value_type tmp{};
        tmp += lhs;
        tmp += rhs;
        return tmp;
      });
```
to avoid requiring for `value_type` to have `operator+`? Otherwise, we should probably make it having `operator+` a requirement for `thrust` to be used.

Reply:
Good catch, I will test this, but looks like a good change to me